File: Other/XML Partner 2.60/examples/Delphi/EXMLPro/ExView.pas
(* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is TurboPower XMLPartner
 *
 * The Initial Developer of the Original Code is
 * TurboPower Software
 *
 * Portions created by the Initial Developer are Copyright (C) 1997-2002
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *
 * ***** END LICENSE BLOCK ***** *)

{*********************************************************}
{* XMLPartner Professional: ExView.PAS 2.57              *}
{*********************************************************}
{* XMLPartner: XSL Preview window for XML editor example *}
{*********************************************************}

unit ExView;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, ExtCtrls, ComCtrls, XpvFlHTM, XpvFlBas, XpvFlRTF, XpvFlPrt,
  XpvXSLPr, XpDOM, XpBase, XpvFlXML, XpvXSLT, ToolWin, ImgList;

type
  { This type describes the interface for a method that implements the
    display of a transformed XML document. }
  TXpFilterDisplayMethod = procedure(aFilter : TXpFilterBase) of object;

  TfrmPreview = class(TForm)
    pnlTop: TPanel;
    pbRender: TButton;
    pbBrowse: TButton;
    cboStylesheet: TComboBox;
    cboFilter: TComboBox;
    lblFilter: TLabel;
    lblStylesheet: TLabel;
    pgPreview: TPageControl;
    tabRTF: TTabSheet;
    tabPrinter: TTabSheet;
    RichEdit: TRichEdit;
    pnlPreviewContainer: TPanel;
    filtPrinter: TXpFilterPrint;
    PrinterHScroll: TScrollBar;
    tabXML: TTabSheet;
    memXML: TMemo;
    tabText: TTabSheet;
    memText: TMemo;
    dlgOpen: TOpenDialog;
    xslProcessor: TXpXSLProcessor;
    filtHTML: TXpFilterHTML;
    filtRTF: TXpFilterRTF;
    filtXML: TXpFilterXML;
    filtText: TXpFilterText;
    tbPrint: TToolBar;
    pbFirstPage: TToolButton;
    pbPrevPage: TToolButton;
    pbNextPage: TToolButton;
    pbLastPage: TToolButton;
    imMain: TImageList;
    lblSpacer: TLabel;
    pnlPage: TPanel;
    lblSpacer2: TLabel;
    PrinterVScroll: TScrollBar;
    procedure FormShow(Sender: TObject);
    procedure cboFilterChange(Sender: TObject);
    procedure FormClose(Sender: TObject; var Action: TCloseAction);
    procedure pbRenderClick(Sender: TObject);
    procedure pbBrowseClick(Sender: TObject);
    procedure pbUpClick(Sender: TObject);
    procedure pbDownClick(Sender: TObject);
    procedure PrinterVScrollChange(Sender: TObject);
    procedure PrinterHScrollChange(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure filtPrinterResize(Sender: TObject);
    procedure pbPrevPageClick(Sender: TObject);
    procedure pbNextPageClick(Sender: TObject);
    procedure pbFirstPageClick(Sender: TObject);
    procedure pbLastPageClick(Sender: TObject);
  private
    { Private declarations }

    FCurSheet : string;
      { Used by SaveStylesheetSelection and RestoreStylesheetSelection. }

    FPrinterFilter : TXpFilterPrint;
      { The printer filter. For use by the printer page controls. }
    FDOM : TXpObjModel;
      { The DOM containing the XML that is to be transformed. }
    FFilename : string;
      { The name of the document being viewed. }
    FLastPage : Integer;
      { The last visible notebook page. }

    procedure DisplayHTML(aFilter : TXpFilterBase);
    procedure DisplayPrinter(aFilter : TXpFilterBase);
    procedure DisplayRTF(aFilter : TXpFilterBase);
    procedure DisplayText(aFilter : TXpFilterBase);
    procedure DisplayXML(aFilter : TXpFilterBase);
    function GetStyleSheet : DOMString;
    procedure RestoreStylesheetSelection;
    procedure SaveStylesheetSelection;
    procedure ShowTab;
    procedure UpdatePrinterUI;
  public
    { Public declarations }

    procedure AddStylesheet(const sName : string; oChildWin : TForm);
      { Add a stylesheet to this window's list of stylesheets. }

    procedure CopyToClipboard;
      { Copies text from the output control currently being viewed. }

    procedure RemoveStylesheet(const sName : string);
      { Remove a stylesheet from this window's list of stylesheets. }

    procedure RenameStylesheet(const sOldName, sNewName : string);
      { Rename a stylesheet in this window's list of stylesheets. }

    property DOM : TXpObjModel read FDOM write FDOM;
      { The DOM containing the XML that is to be transformed. }

    property Filename : string read FFilename write FFilename;
      { The name of the XML document being viewed. }

  end;

var
  frmPreview: TfrmPreview;

implementation

uses
{$IFDEF WIN32}
  ShellAPI,
{$ENDIF}
  ExChildw,
  ExErr;

{$R *.DFM}

const
  ciTabNone = -1;
  ciTabRTF = 0;
  ciTabPrinter = 1;
  ciTabXML = 2;
  ciTabText = 3;

type
  { This type is used to store filter-specific data. One instance of this
    class is associated with each entry in the filter combobox. }
  TXpFilterDetail = class
  public
    Display : TXpFilterDisplayMethod;
      { The method used to display the rendered XML document. }
    Filter : TXpFilterBase;
      { The filter component to be used to render the XML document. }
    TabIndex : Integer;
      { The notebook tab associated with the filter. If set to ciTabNone then
        the filter does not have a tab. }
  end;

procedure TfrmPreview.FormShow(Sender: TObject);
var
  aFilter : TXpFilterBase;
  aFilterDetail : TXpFilterDetail;
  Inx : Integer;
begin
  FLastPage := ciTabNone;
  { Find the filters on the form & populate the Filter combobox. }
  for Inx := 0 to pred(ComponentCount) do begin
    if Components[Inx] is TXpFilterBase then begin
      aFilter := TXpFilterBase(Components[Inx]);
      aFilterDetail := TXpFilterDetail.Create;
      aFilterDetail.Filter  := aFilter;
      if Components[Inx].ClassType = TXpFilterHTML then begin
        aFilterDetail.Display := DisplayHTML;
        aFilterDetail.TabIndex := ciTabNone;
      end
      else if Components[Inx].ClassType = TXpFilterPrint then begin
        aFilterDetail.Display := DisplayPrinter;
        aFilterDetail.TabIndex := ciTabPrinter;
      end
      else if Components[Inx].ClassType = TXpFilterRTF then begin
        aFilterDetail.Display := DisplayRTF;
        aFilterDetail.TabIndex := ciTabRTF;
      end
      else if Components[Inx].ClassType = TXpFilterXML then begin
        aFilterDetail.Display := DisplayXML;
        aFilterDetail.TabIndex := ciTabXML;
      end
      else begin
        aFilterDetail.Display := DisplayText;
        aFilterDetail.TabIndex := ciTabText;
      end;
      cboFilter.Items.AddObject(aFilter.DisplayName, Pointer(aFilterDetail));
    end;
  end;

  { Position to the first filter. }
  if cboFilter.Items.Count > 0 then
    cboFilter.ItemIndex := 0
  else begin
    cboFilter.Items.Add('<None available>');
    cboFilter.ItemIndex := 0;
    cboFilter.Enabled := False;
  end;

  { Position to the first stylesheet. }
  if cboStyleSheet.Items.Count > 0 then
    cboStyleSheet.ItemIndex := 0;

  { Set to the appropriate notebook tab. }
  ShowTab;

end;

procedure TfrmPreview.cboFilterChange(Sender: TObject);
begin
  { Display the tab associated with the selected filter. }
  ShowTab;
end;

procedure TfrmPreview.CopyToClipboard;
begin
  if pgPreview.ActivePage = tabRTF then
    RichEdit.CopyToClipboard
  else if pgPreview.ActivePage = tabXML then
    memXML.CopyToClipboard
  else if pgPreview.ActivePage = tabText then
    memText.CopyToClipboard;
end;

procedure TfrmPreview.FormClose(Sender: TObject; var Action: TCloseAction);
var
  Inx : Integer;
begin
  { Free the filter detail objects associated with the items in the combobox. }
  for Inx := pred(cboFilter.Items.Count) downto 0 do
    cboFilter.Items.Objects[Inx].Free;
end;

procedure TfrmPreview.ShowTab;
var
  aFilterDetail : TXpFilterDetail;
begin

  aFilterDetail := TXpFilterDetail(cboFilter.Items.Objects[cboFilter.ItemIndex]);

  if FLastPage = aFilterDetail.TabIndex then
    Exit;

  { Hide the previously shown page. }
  if FLastPage <> ciTabNone then
    pgPreview.Pages[FLastPage].TabVisible := False;

  FLastPage := aFilterDetail.TabIndex;
  if FLastPage <> ciTabNone then begin
    pgPreview.ActivePage := pgPreview.Pages[FLastPage];
    pgPreview.Pages[FLastPage].TabVisible := True;
    pgPreview.Pages[FLastPage].SetFocus;
    if FLastPage = ciTabPrinter then
      UpdatePrinterUI;
  end;
end;

procedure TfrmPreview.pbRenderClick(Sender: TObject);
var
  aFilterDetail : TXpFilterDetail;
  aFilter : TXpFilterBase;
  savCursor : TCursor;
begin
  { Get the filter. }
  aFilterDetail := TXpFilterDetail(cboFilter.Items.Objects[cboFilter.ItemIndex]);
  aFilter := TXpFilterBase(aFilterDetail.Filter);
  with xslProcessor do begin
    Filter := aFilter;
    XMLObjModel := FDOM;
    { Did the user specify an external stylesheet? }
    if cboStyleSheet.ItemIndex < 0 then begin
      { Yes. Get the name of the file. }
      if cboStylesheet.Text <> '' then
        StyleURL := cboStyleSheet.Text;
    end
    else
      { No. Obtain the data for the stylesheet. }
      StyleData := GetStyleSheet;

    savCursor := Screen.Cursor;
    Screen.Cursor := crHourGlass;
    try
      if ApplyStyle then
        { Display the file using the method specific to the filter. }
        aFilterDetail.Display(aFilter)
      else begin
        Screen.Cursor := savCursor;
        with TfrmErrors.Create(Self) do
          try
            Errors := xslProcessor.Errors;
            Caption := 'XSL Errors';
            IntroMsg := 'applying the stylesheet';
            ShowModal;
          finally
            Free;
          end;
      end;
    finally
      Screen.Cursor := savCursor;
    end;
  end;
end;

function TfrmPreview.GetStyleSheet : DOMString;
var
  bIsBEorLE : Boolean;
  oForm : TXmlChild;
  sXML : DOMString;
begin
  oForm := TXmlChild(cboStyleSheet.Items.Objects[cboStyleSheet.ItemIndex]);
  sXML := oForm.XMLData;
  bIsBEorLE := ( (sXML[1] = #$FEFF) or
                 (sXML[1] = #$FFFE) );
  { Marked as big or little endian? }
  if not bIsBEorLE then begin
   { No. Must force it to be recognized as little endian. }
   Result := DOMString(#$FEFF) + sXML;
  end
  else
    Result := sXML;
end;

procedure TfrmPreview.DisplayHTML(aFilter : TXpFilterBase);
resourcestring
  EX_Error = 'Unable to start web browser. Make sure you have it properly ' +
             'set up on your system.';
var
  sFile : string;
begin
  if FFileName = '' then
    FFileName := 'temp';
  sFile := XpMakeFilePath(ExtractFilePath(Application.ExeName),
                          ChangeFileExt(ExtractFileName(FFileName),'.htm'));
  TXpFilterHTML(aFilter).OutputEncoding := FDOM.OutCharSet;
  aFilter.SaveToFile(sFile);
{$IFDEF WIN32}
  if ShellExecute(0, 'open', PChar(sFile), '', '', SW_SHOWNORMAL) <= 32 then
    ShowMessage(EX_Error);
{$ENDIF}

{$IFDEF LINUX}
{$ENDIF}
end;

procedure TfrmPreview.DisplayPrinter(aFilter : TXpFilterBase);
begin
  FPrinterFilter := TXpFilterPrint(aFilter);
  UpdatePrinterUI;
end;

procedure TfrmPreview.DisplayRTF(aFilter : TXpFilterBase);
var
  sFile : string;
begin
  sFile := XpMakeFilePath(ExtractFilePath(Application.ExeName),
                          ChangeFileExt(ExtractFileName(FFileName),'.rtf'));
  aFilter.SaveToFile(sFile);
  RichEdit.Clear;
  RichEdit.Lines.LoadFromFile(sFile);
end;

procedure TfrmPreview.DisplayText(aFilter : TXpFilterBase);
var
  sFile : string;
begin
  sFile := XpMakeFilePath(ExtractFilePath(Application.ExeName),
                          ChangeFileExt(ExtractFileName(FFileName),'.txt'));
  TXpFilterText(aFilter).OutputEncoding := FDOM.OutCharSet;
  aFilter.SaveToFile(sFile);
  memText.Lines.Clear;
  memText.Lines.LoadFromFile(sFile);
end;

procedure TfrmPreview.DisplayXML(aFilter : TXpFilterBase);
var
  sFile : string;
begin
  sFile := XpMakeFilePath(ExtractFilePath(Application.ExeName),
                          ChangeFileExt(ExtractFileName(FFileName),'.out'));
  TXpFilterXML(aFilter).OutputEncoding := FDOM.OutCharSet;
  aFilter.SaveToFile(sFile);
  memXML.Lines.Clear;
  memXML.Lines.LoadFromFile(sFile);
end;

procedure TfrmPreview.pbBrowseClick(Sender: TObject);
begin
  if dlgOpen.Execute then begin
    cboStylesheet.ItemIndex := -1;
    cboStylesheet.Text := dlgOpen.FileName;
  end;
end;

procedure TfrmPreview.pbUpClick(Sender: TObject);
begin
  if FPrinterFilter.PageCurrent <= 1 then begin
    MessageBeep(MB_OK);
    exit;
  end;

  FPrinterFilter.PageCurrent := FPrinterFilter.PageCurrent - 1;
  UpdatePrinterUI;
end;

procedure TfrmPreview.pbDownClick(Sender: TObject);
begin
  if FPrinterFilter.PageCurrent >= FPrinterFilter.PageCount then begin
    MessageBeep(MB_OK);
    exit;
  end;

  FPrinterFilter.PageCurrent := FPrinterFilter.PageCurrent + 1;
  UpdatePrinterUI;
end;

procedure TfrmPreview.UpdatePrinterUI;
var
  bFirstPage,
  bHasPages,
  bLastPage : Boolean;
begin
  if Assigned(FPrinterFilter) then begin
    pnlPage.Caption := Format('Page %d of %d',
                                   [FPrinterFilter.PageCurrent,
                                    FPrinterFilter.PageCount]);
    bHasPages := (FPrinterFilter.PageCount > 0);
  end
  else
    bHasPages := False;

  if bHasPages then begin
    PrinterVScroll.Position := 0;
    PrinterHScroll.Position := 0;
    bFirstPage := (FPrinterFilter.PageCurrent = 1);
    bLastPage := (FPrinterFilter.PageCurrent = FPrinterFilter.PageCount);
    pbFirstPage.Enabled := (not bFirstPage);
    pbPrevPage.Enabled := (not bFirstPage);
    pbNextPage.Enabled := (not bLastPage);
    pbLastPage.Enabled := (not bLastPage);
  end
  else begin
    pbFirstPage.Enabled := False;
    pbPrevPage.Enabled := False;
    pbNextPage.Enabled := False;
    pbLastPage.Enabled := False;
  end;
end;

procedure TfrmPreview.PrinterVScrollChange(Sender: TObject);
begin
  FPrinterFilter.VSlidePosition := PrinterVScroll.Position;
end;

procedure TfrmPreview.PrinterHScrollChange(Sender: TObject);
begin
  FPrinterFilter.HSlidePosition := PrinterHScroll.Position;
end;

procedure TfrmPreview.FormCreate(Sender: TObject);
var
  Inx : Integer;
begin
  { Make sure the tab pages are invisible. }
  for Inx := 0 to pred(pgPreview.PageCount) do
    pgPreview.Pages[Inx].TabVisible := False;
end;

procedure TfrmPreview.filtPrinterResize(Sender : TObject);
begin
  PrinterVScroll.Max := filtPrinter.VSlideFactor;
  PrinterVScroll.Position := filtPrinter.VSlidePosition;

  PrinterHScroll.Max := filtPrinter.HSlideFactor;
  PrinterHScroll.Position := filtPrinter.HSlidePosition;
end;

procedure TfrmPreview.SaveStylesheetSelection;
begin
  if cboStylesheet.ItemIndex >= 0 then
    FCurSheet := cboStylesheet.Items[cboStylesheet.ItemIndex]
  else
    FCurSheet := '';
end;

procedure TfrmPreview.RestoreStylesheetSelection;
var
  wInx : Integer;
begin
  if FCurSheet <> '' then begin
    wInx := cboStylesheet.Items.IndexOf(FCurSheet);
    { Is the stylesheet still in the list? }
    if wInx >= 0 then
      { Yes. Position combobox to that stylesheet. }
      cboStylesheet.ItemIndex := wInx
    else begin
      { No. Clear out the combobobox & the XSL Processor's style data. }
      cboStylesheet.Text := '';
      xslProcessor.StyleData := '';
    end;
  end;
end;

procedure TfrmPreview.AddStylesheet(const sName : string; oChildWin : TForm);
begin
  SaveStylesheetSelection;
  try
    cboStylesheet.Items.AddObject(sName, oChildWin);
  finally
    RestoreStylesheetSelection;
  end;
end;

procedure TfrmPreview.RemoveStylesheet(const sName : string);
var
  wInx : Integer;
begin
  SaveStylesheetSelection;
  try
    wInx := cboStylesheet.Items.IndexOf(sName);
    if wInx >= 0 then
      cboStylesheet.Items.Delete(wInx);
  finally
    RestoreStylesheetSelection;
  end;
end;

procedure TfrmPreview.RenameStylesheet(const sOldName, sNewName : string);
var
  wInx : Integer;
begin
  SaveStylesheetSelection;
  try
    if FCurSheet = sOldName then
      FCurSheet := sNewName;
    wInx := cboStylesheet.Items.IndexOf(sOldName);
    if wInx >= 0 then
      cboStylesheet.Items[wInx] := sNewName;
  finally
    RestoreStylesheetSelection;
  end;
end;

procedure TfrmPreview.pbPrevPageClick(Sender: TObject);
begin
  if FPrinterFilter.PageCurrent <= 1 then begin
    MessageBeep(MB_OK);
    exit;
  end;

  FPrinterFilter.PageCurrent := FPrinterFilter.PageCurrent - 1;
  UpdatePrinterUI;
end;

procedure TfrmPreview.pbNextPageClick(Sender: TObject);
begin
  if FPrinterFilter.PageCurrent >= FPrinterFilter.PageCount then begin
    MessageBeep(MB_OK);
    exit;
  end;

  FPrinterFilter.PageCurrent := FPrinterFilter.PageCurrent + 1;
  UpdatePrinterUI;
end;

procedure TfrmPreview.pbFirstPageClick(Sender: TObject);
begin
  FPrinterFilter.PageCurrent := 1;
  UpdatePrinterUI;
end;

procedure TfrmPreview.pbLastPageClick(Sender: TObject);
begin
  FPrinterFilter.PageCurrent := FPrinterFilter.PageCount;
  UpdatePrinterUI;
end;
end.

File: Other/XML Partner 2.60/examples/CBuildr5/EXMLPro/ExView.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#pragma hdrstop

#include "ExView.h"
#include "ExChildW.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "XpBase"
#pragma link "XpvFlBas"
#pragma link "XpvFlHTM"
#pragma link "XpvFlPrt"
#pragma link "XpvFlRTF"
#pragma link "XpvFlXML"
#pragma link "XpvXSLPr"
#pragma link "XpvXSLT"
#pragma resource "*.dfm"
//---------------------------------------------------------------------------
const ciTabNone = -1;
const ciTabRTF = 0;
const ciTabPrinter = 1;
const ciTabXML = 2;
const ciTabText = 3;
//---------------------------------------------------------------------------
TfrmPreview *frmPreview;
//---------------------------------------------------------------------------
__fastcall TfrmPreview::TfrmPreview(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------
class TXpFilterDetail : public TObject
  // This class is used to store filter-specific data. One instance of this
  // class is associated with each entry in the filter combobox.
{
public:
  TXpFilterDisplayMethod Display;
    // The method used to display the rendered XML document.

  TXpFilterBase* Filter;
    // The filter component to be used to render the XML document.

  int TabIndex;
    // The notebook tab associated with the filter. If set to ciTabNone then
    // the filter does not have a tab.
};
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::DisplayHTML(TXpFilterBase* aFilter)
{
  String EX_Error = "Unable to start web browser. Make sure you have it properly set up on your system.";

  String sFile;

  if (FFilename == "")
    FFilename = "temp";

  sFile = XpMakeFilePath(ExtractFilePath(Application->ExeName),
                         ChangeFileExt(ExtractFileName(FFilename),".htm"));
  dynamic_cast<TXpFilterHTML*>(aFilter)->OutputEncoding = FDOM->OutCharSet;
  aFilter->SaveToFile(sFile);
  char* tmpFilename = sFile.c_str();
  if (int(ShellExecute(0, "open", tmpFilename, "", "", SW_SHOWNORMAL)) <= 32)
    ShowMessage(EX_Error);
}
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::DisplayPrinter(TXpFilterBase* aFilter)
{
  FPrinterFilter = dynamic_cast<TXpFilterPrint*>(aFilter);
  UpdatePrinterUI();
}
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::DisplayRTF(TXpFilterBase* aFilter)
{
  String sFile;

  sFile = XpMakeFilePath(ExtractFilePath(Application->ExeName),
                         ChangeFileExt(ExtractFileName(FFilename),".rtf"));
  aFilter->SaveToFile(sFile);
  RichEdit->Clear();
  RichEdit->Lines->LoadFromFile(sFile);
}
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::DisplayText(TXpFilterBase* aFilter)
{
  String sFile;

  sFile = XpMakeFilePath(ExtractFilePath(Application->ExeName),
                         ChangeFileExt(ExtractFileName(FFilename),".txt"));
  dynamic_cast<TXpFilterText*>(aFilter)->OutputEncoding = FDOM->OutCharSet;
  aFilter->SaveToFile(sFile);
  memText->Lines->Clear();
  memText->Lines->LoadFromFile(sFile);
}
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::DisplayXML(TXpFilterBase* aFilter)
{
  String sFile;

  sFile = XpMakeFilePath(ExtractFilePath(Application->ExeName),
                         ChangeFileExt(ExtractFileName(FFilename),".xml"));
  dynamic_cast<TXpFilterXML*>(aFilter)->OutputEncoding = FDOM->OutCharSet;
  aFilter->SaveToFile(sFile);
  memXML->Lines->Clear();
  memXML->Lines->LoadFromFile(sFile);
}
//---------------------------------------------------------------------------
WideString __fastcall TfrmPreview::GetStyleSheet(void)
{
  bool bIsBEorLE;
  WideString sXML;
  WideString sXMLBOM;

  sXML = dynamic_cast<TXMLChild*>(cboStylesheet->Items->Objects[cboStylesheet->ItemIndex])->XMLData;
  bIsBEorLE = ( ((sXML[1] == 254) & (sXML[2] == 257)) |
                ((sXML[1] == 257) & (sXML[2] == 254)) );
  // Marked as big or little endian?
  if (!bIsBEorLE)
  {
    // No. Must force it to be recognized as little endian.
    sXMLBOM.SetLength(1);
    sXMLBOM[1] = 65279;  // $FEFF;
    return(sXMLBOM + sXML);
  }
  else
    return(sXML);
}
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::RestoreStylesheetSelection(void)
{
  int wInx;

  if (FCurSheet != "")
  {
    wInx = cboStylesheet->Items->IndexOf(FCurSheet);
    // Is the stylesheet still in the list?
    if (wInx >= 0)
      // Yes. Position combobox to that stylesheet.
      cboStylesheet->ItemIndex = wInx;
    else
    {
      // No. Clear out the combobobox & the XSL Processor's style data.
      cboStylesheet->Text = "";
      xslprocessor->StyleData = "";
    }
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::SaveStylesheetSelection(void)
{
  if (cboStylesheet->ItemIndex >= 0)
    FCurSheet = cboStylesheet->Items->Strings[cboStylesheet->ItemIndex];
  else
    FCurSheet = "";
}
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::ShowTab(void)
{
  TXpFilterDetail* aFilterDetail;

  aFilterDetail = dynamic_cast<TXpFilterDetail*>(cboFilter->Items->Objects[cboFilter->ItemIndex]);

  if (FLastPage == aFilterDetail->TabIndex)
    return;

  // Hide the previously shown page.
  if (FLastPage != ciTabNone)
    pgPreview->Pages[FLastPage]->TabVisible = false;

  FLastPage = aFilterDetail->TabIndex;
  if (FLastPage != ciTabNone)
  {
    pgPreview->ActivePage = pgPreview->Pages[FLastPage];
    pgPreview->Pages[FLastPage]->TabVisible = true;
    pgPreview->Pages[FLastPage]->SetFocus();
    if (FLastPage == ciTabPrinter)
      UpdatePrinterUI();
  }

}
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::UpdatePrinterUI(void)
{
  bool bFirstPage;
  bool bHasPages;
  bool bLastPage;

  if (FPrinterFilter != NULL)
  {
    pnlPage->Caption = Format("Page %d of %d",
                              ARRAYOFCONST((FPrinterFilter->PageCurrent,
                                            FPrinterFilter->PageCount)));
    bHasPages = (FPrinterFilter->PageCount > 0);
  }
  else
    bHasPages = false;

  if (bHasPages)
  {
    PrinterVScroll->Position = 0;
    PrinterHScroll->Position = 0;
    bFirstPage = (FPrinterFilter->PageCurrent == 1);
    bLastPage = (FPrinterFilter->PageCurrent == FPrinterFilter->PageCount);
    pbFirstPage->Enabled = (!bFirstPage);
    pbPrevPage->Enabled = (!bFirstPage);
    pbNextPage->Enabled = (!bLastPage);
    pbLastPage->Enabled = (!bLastPage);
  }
  else
  {
    pbFirstPage->Enabled = False;
    pbPrevPage->Enabled = False;
    pbNextPage->Enabled = False;
    pbLastPage->Enabled = False;
  }
}
//---------------------------------------------------------------------------
//---------------------------------------------------------------------------
//---------------------------------------------------------------------------
//---------------------------------------------------------------------------
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::cboFilterChange(TObject *Sender)
{
  // Display the tab associated with the selected filter.
  ShowTab();
}
//---------------------------------------------------------------------------

void __fastcall TfrmPreview::pbBrowseClick(TObject *Sender)
{
  if (dlgOpen->Execute())
  {
    cboStylesheet->ItemIndex = -1;
    cboStylesheet->Text = dlgOpen->FileName;
  }
}
//---------------------------------------------------------------------------

void __fastcall TfrmPreview::pbRenderClick(TObject *Sender)
{
  TXpFilterDetail *aFilterDetail;
  TXpFilterBase *aFilter;
  TCursor savCursor;

  // Get the filter.
  aFilterDetail = dynamic_cast<TXpFilterDetail*>(cboFilter->Items->Objects[cboFilter->ItemIndex]);
  aFilter = dynamic_cast<TXpFilterBase*>(aFilterDetail->Filter);
  xslprocessor->Filter = aFilter;
  xslprocessor->XmlObjModel = FDOM;
  // Did the user specify an external stylesheet?
  if (cboStylesheet->ItemIndex < 0)
  {
    // Yes. Get the name of the file.
    if (cboStylesheet->Text != "")
      xslprocessor->StyleURL = cboStylesheet->Text;
  }
  else
    // No. Obtain the data for the stylesheet.
    xslprocessor->StyleData = GetStyleSheet();

  savCursor = Screen->Cursor;
  Screen->Cursor = crHourGlass;
  try
  {
    if (xslprocessor->ApplyStyle())
      // Display the file using the method specific to the filter.
      aFilterDetail->Display(aFilter);
    else
    {
      Screen->Cursor = savCursor;
      MessageDlg("Render failed!", mtError, TMsgDlgButtons() << mbOK, 0);
    }
  }
  __finally
  {
    Screen->Cursor = savCursor;
  }
}
//---------------------------------------------------------------------------

void __fastcall TfrmPreview::filtPrinterResize(TObject *Sender)
{
  PrinterVScroll->Max = filtPrinter->VSlideFactor;
  PrinterVScroll->Position = filtPrinter->VSlidePosition;

  PrinterHScroll->Max = filtPrinter->HSlideFactor;
  PrinterHScroll->Position = filtPrinter->HSlidePosition;
}
//---------------------------------------------------------------------------

void __fastcall TfrmPreview::PrinterVScrollChange(TObject *Sender)
{
  FPrinterFilter->VSlidePosition = PrinterVScroll->Position;
}
//---------------------------------------------------------------------------

void __fastcall TfrmPreview::PrinterHScrollChange(TObject *Sender)
{
  FPrinterFilter->HSlidePosition = PrinterHScroll->Position;
}
//---------------------------------------------------------------------------

void __fastcall TfrmPreview::pbFirstPageClick(TObject *Sender)
{
  FPrinterFilter->PageCurrent = 1;
  UpdatePrinterUI();
}
//---------------------------------------------------------------------------

void __fastcall TfrmPreview::pbPrevPageClick(TObject *Sender)
{
  if (FPrinterFilter->PageCurrent <= 1)
  {
    MessageBeep(MB_OK);
    return;
  }

  FPrinterFilter->PageCurrent = FPrinterFilter->PageCurrent - 1;
  UpdatePrinterUI();

}
//---------------------------------------------------------------------------

void __fastcall TfrmPreview::pbNextPageClick(TObject *Sender)
{
  if (FPrinterFilter->PageCurrent >= FPrinterFilter->PageCount)
  {
    MessageBeep(MB_OK);
    return;
  }

  FPrinterFilter->PageCurrent = FPrinterFilter->PageCurrent + 1;
  UpdatePrinterUI();

}
//---------------------------------------------------------------------------

void __fastcall TfrmPreview::pbLastPageClick(TObject *Sender)
{
  FPrinterFilter->PageCurrent = FPrinterFilter->PageCount;
  UpdatePrinterUI();
}
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::AddStylesheet(const String sName, TForm* oChildWin)
{
  SaveStylesheetSelection();
  try
  {
    cboStylesheet->Items->AddObject(sName, oChildWin);
  }
  __finally
  {
    RestoreStylesheetSelection();
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::CopyToClipboard(void)
{
  if (pgPreview->ActivePage == tabRTF)
    RichEdit->CopyToClipboard();
  else if (pgPreview->ActivePage == tabXML)
    memXML->CopyToClipboard();
  else if (pgPreview->ActivePage == tabText)
    memText->CopyToClipboard();
}
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::RemoveStylesheet(const String sName)
{
  int wInx;

  SaveStylesheetSelection();
  try
  {
    wInx = cboStylesheet->Items->IndexOf(sName);
    if (wInx >= 0)
      cboStylesheet->Items->Delete(wInx);
  }
  __finally
  {
    RestoreStylesheetSelection();
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmPreview::RenameStylesheet(const String sOldName, const String sNewName)
{
  int wInx;

  SaveStylesheetSelection();
  try
  {
    if (FCurSheet == sOldName)
      FCurSheet = sNewName;
    wInx = cboStylesheet->Items->IndexOf(sOldName);
    if (wInx >= 0)
      cboStylesheet->Items->Strings[wInx] = sNewName;
  }
  __finally
  {
    RestoreStylesheetSelection();
  }
}
//---------------------------------------------------------------------------

void __fastcall TfrmPreview::FormShow(TObject *Sender)
{
  TXpFilterBase* aFilter;
  TXpFilterDetail* aFilterDetail;
  int Inx;

  FLastPage = ciTabNone;
  // Find the filters on the form & populate the Filter combobox.
  for (Inx = 0; Inx < ComponentCount; Inx++)
  {
    aFilter = dynamic_cast<TXpFilterBase*>(Components[Inx]);
    if (aFilter != NULL)
    {
      aFilterDetail = new TXpFilterDetail();
      aFilterDetail->Filter = aFilter;
      if (Components[Inx]->ClassType() == __classid(TXpFilterHTML))
      {
        aFilterDetail->Display = DisplayHTML;
        aFilterDetail->TabIndex = ciTabNone;
      }
      else if (Components[Inx]->ClassType() == __classid(TXpFilterPrint))
      {
        aFilterDetail->Display = DisplayPrinter;
        aFilterDetail->TabIndex = ciTabPrinter;
      }
      else if (Components[Inx]->ClassType() == __classid(TXpFilterRTF))
      {
        aFilterDetail->Display = DisplayRTF;
        aFilterDetail->TabIndex = ciTabRTF;
      }
      else if (Components[Inx]->ClassType() == __classid(TXpFilterXML))
      {
        aFilterDetail->Display = DisplayXML;
        aFilterDetail->TabIndex = ciTabXML;
      }
      else
      {
        aFilterDetail->Display = DisplayText;
        aFilterDetail->TabIndex = ciTabText;
      }
      cboFilter->Items->AddObject(aFilter->DisplayName, aFilterDetail);
    }
  }

  // Position to the first filter.
  if (cboFilter->Items->Count > 0)
    cboFilter->ItemIndex = 0;
  else
  {
    cboFilter->Items->Add("<None available>");
    cboFilter->ItemIndex = 0;
    cboFilter->Enabled = false;
  }

  // Position to the first stylesheet.
  if (cboStylesheet->Items->Count > 0)
    cboStylesheet->ItemIndex = 0;

  // Set to the appropriate notebook tab.
  ShowTab();
}
//---------------------------------------------------------------------------

void __fastcall TfrmPreview::FormCreate(TObject *Sender)
{
  int Inx;

  // Make sure the tab pages are invisible.
  for (Inx = 0; Inx < pgPreview->PageCount; Inx++)
    pgPreview->Pages[Inx]->TabVisible = false;

}
//---------------------------------------------------------------------------

void __fastcall TfrmPreview::FormClose(TObject *Sender,
      TCloseAction &Action)
{
  int Inx;

  // Free the filter detail objects associated with the items in the combobox.
  for (Inx = cboFilter->Items->Count - 1; Inx >= 0; Inx--)
    cboFilter->Items->Objects[Inx]->Free();

}
//---------------------------------------------------------------------------

